### DISCUSSION TITLE
Code organization

### DISCUSSION THREAD

**Discussion by onlyjus** [ESMS-Group-Public/FoKL-GPy#5] [General]
Why are there copies (didn't check if the code is identical, but the doc strings are different) of all the routines in independent files and collected all under the FoKLRoutines.py:FoKL class? Which code is correct? The Lorenz and Sinusoid examples both us the FoKL class.

Also, it would be nice to mimic scikit-learn calls.

```python
# import
from FoKL import FoKL

# construct model, set defaults so you don't have to specify the same values over and over again
model = FoKL(relats_in, a, b, atau, btau, tolerance, draws, gimmie, way3, threshav, threshstda, threshstdb, aic)

# fit the model
model.fit(X, y)  # could use .train(X, y)

# predict
mean, bounds = model.predict(Xnew, bounds=True)
```

**Comment by dmebane** [comment]
Hi Justin thanks for those comments. We are a pretty lean outfit and don't have a lot of resources or knowledge on the user experience side, but we are indeed making some changes to make things a little easier / more familiar. Should be pushing some changes soon!

**Reply by onlyjus** [reply]
Also, can the phis be computed instead of reading in the text files? That way you could generalize your function calls instead of having  these `499/498`s all over the place.

**Reply by dmebane** [reply]
Hey Justin -- the basis functions are nonparametric (as will be the case for any GP -- this is a big advantage in fact since most experimental data is also nonparametric) and as such we represent them with splines. Do those numbers come into any of the function calls? (Sorry for being not completely familiar with the python version of the package.)

**Comment by derek-slack** [comment]
Hey Justin. We just pushed FoKL 2.0.1, it brings a new calling structure similar to the one you outlined. The repos read me is updated as well to show this new methodology. I appreciate your input on this and hope to bring more QoL changes to make FoKL more accessible. Let me know if you have any more questions.

**Reply by onlyjus** [reply]
Awesome! That is much cleaner. The other thing I would suggest is to set defaults for all "hyper parameters" in the `__init__`:

`def __init__(self, phis, relats_in=[], a=9, b=0.01, atau=3, btau=4000, tolerance=3, draws=1000, gimmie=False, way3=False, threshav=0.05, threshstda=0.5, threshstdb=2, aic=True):`

That way you don't have to keep repeating values over and over again and you give the user a reference for a recommended value.

IMO if you want people to use the code you write, I have found that you really need to reduce the "barrier to entry" by providing default values, simple intuitive calls, etc. Once they try it and it works, then they will dive into reading/playing with parameters.

**Reply by dmebane** [reply]
Hi Justin -- this is absolutely correct: useability / user experience issues are vitally important. We are however a pretty small operation, and so it might take us some time to make all the changes we need to. In the meantime the main thing to know is that the hyperparameters a, b, atau and btau will change depending on the dataset: a and atau can stay at 4, but b and btau then need to be scaled using features of the data: b = 5*sigma^2, where sigma^2 is the best prior estimate of the variance of the observation error in the data. btau is then 5*(tau^2)/(sigma^2), where tau^2 is the variance of the dataset itself.

Thanks for trying it out and for providing helpful suggestions!!
